File: backend/src/main/java/com/example/Services/LibraryInvoicePdfService.java
package com.example.Services;

import com.example.models.LibraryPackagePurchase;
import com.example.models.LibraryPackagePurchaseItem;
import com.example.models.Product;
import com.lowagie.text.Document;
import com.lowagie.text.Font;
import com.lowagie.text.FontFactory;
import com.lowagie.text.PageSize;
import com.lowagie.text.Paragraph;
import com.lowagie.text.pdf.PdfPTable;
import com.lowagie.text.pdf.PdfWriter;
import org.springframework.stereotype.Service;

import java.io.ByteArrayOutputStream;
import java.util.List;

/** Invoice for a library package purchase and the titles borrowed on it. */
@Service
public class LibraryInvoicePdfService {

    public byte[] generateLibraryInvoice(LibraryPackagePurchase purchase,
                                         List<LibraryPackagePurchaseItem> items) {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        Document document = new Document(PageSize.A4);
        PdfWriter.getInstance(document, out);
        document.open();

        Font titleFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 18);
        Font normalFont = FontFactory.getFont(FontFactory.HELVETICA, 12);
        Font boldFont = FontFactory.getFont(FontFactory.HELVETICA_BOLD, 12);

        document.add(new Paragraph("Bookworm - Library Package Invoice", titleFont));
        document.add(new Paragraph(" "));

        document.add(new Paragraph("Invoice ID: LIB-" + purchase.getPurchaseId(), normalFont));
        document.add(new Paragraph(
                "Transaction ID: " + purchase.getTransaction().getTransactionId(), normalFont));
        document.add(new Paragraph("User: " + purchase.getUser().getUserName(), normalFont));
        document.add(new Paragraph("Purchase Date: " + purchase.getPurchaseDate(), normalFont));
        document.add(new Paragraph(" "));

        document.add(new Paragraph("Package Details", boldFont));
        document.add(new Paragraph(
                "Package Name: " + purchase.getLibraryPackage().getName(), normalFont));
        document.add(new Paragraph(
                "Validity: " + purchase.getLibraryPackage().getValidityDays() + " days", normalFont));
        document.add(new Paragraph("Book Limit: " + purchase.getAllowedBooks(), normalFont));
        document.add(new Paragraph(" "));

        PdfPTable table = new PdfPTable(3);
        table.setWidthPercentage(100);
        table.addCell("Book");
        table.addCell("Avg Price");
        table.addCell("Royalty");

        for (LibraryPackagePurchaseItem item : items) {
            table.addCell(invoiceName(item.getProduct()));
            table.addCell("Rs " + purchase.getAvgBookPrice());
            table.addCell("Rs " + item.getRoyaltyAmount());
        }
        document.add(table);

        document.add(new Paragraph(" "));
        document.add(new Paragraph("Total Paid: Rs " + purchase.getPackagePrice(), titleFont));

        document.close();
        return out.toByteArray();
    }

    /**
     * OpenPDF's default font (Helvetica) only has Latin-script glyphs -
     * a title written in Marathi/Hindi/Konkani would print as a blank
     * cell, not an error, so this is easy to miss without knowing to
     * look for it. Falling back to the English name sidesteps that
     * entirely instead of needing to embed a Devanagari font.
     */
    private String invoiceName(Product product) {
        return product.getProductNameEnglish() != null && !product.getProductNameEnglish().isBlank()
                ? product.getProductNameEnglish()
                : product.getProductName();
    }
}

File: dotnet-backend/BookwormApi/Service/LibraryInvoicePdfService.cs
using BookwormApi.Models;
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;

namespace BookwormApi.Service;

public class LibraryInvoicePdfService : ILibraryInvoicePdfService
{
    public byte[] GenerateLibraryInvoice(LibraryPackagePurchase purchase, List<LibraryPackagePurchaseItem> items)
    {
        var document = Document.Create(container =>
        {
            container.Page(page =>
            {
                page.Size(PageSizes.A4);
                page.Margin(30);
                page.DefaultTextStyle(x => x.FontSize(11));

                page.Header().Column(col =>
                {
                    col.Item().Text("Bookworm - Library Package Invoice").FontSize(18).Bold();
                    col.Item().Text($"Invoice ID: LIB-{purchase.PurchaseId}");
                    col.Item().Text($"Transaction ID: {purchase.TransactionId}");
                    col.Item().Text($"User: {purchase.User?.UserName}");
                    col.Item().Text($"Purchase Date: {purchase.PurchaseDate}");
                    col.Item().Text($"Package: {purchase.LibraryPackage?.Name}");
                    col.Item().Text($"Validity: {purchase.LibraryPackage?.ValidityDays} days");
                    col.Item().Text($"Book Limit: {purchase.AllowedBooks}");
                });

                page.Content().PaddingTop(15).Table(table =>
                {
                    table.ColumnsDefinition(columns =>
                    {
                        columns.RelativeColumn(3);
                        columns.RelativeColumn(1);
                        columns.RelativeColumn(1);
                    });

                    table.Header(header =>
                    {
                        header.Cell().Text("Book").Bold();
                        header.Cell().Text("Avg Price").Bold();
                        header.Cell().Text("Royalty").Bold();
                    });

                    foreach (var item in items)
                    {
                        table.Cell().Text(InvoiceNameHelper.InvoiceName(item.Product));
                        table.Cell().Text($"Rs {purchase.AvgBookPrice}");
                        table.Cell().Text($"Rs {item.RoyaltyAmount}");
                    }
                });

                page.Footer().PaddingTop(15).Text($"Total Paid: Rs {purchase.PackagePrice}").Bold();
            });
        });

        return document.GeneratePdf();
    }
}
